.gitignore

```diff
@@ -30,3 +30,32 @@
 *.exe
 *.out
 *.app
+
+# Build Files
+**/CMakeFiles/
+**/CMakeCache.txt
+**cmake_install.cmake
+compile_commands.json
+Makefile
+**CTestTestfile.cmake
+test/**.cmake
+/cmake-build-debug/
```

Comment:
These aren't necessary. In-source builds are never a good idea and not ignoring the files gives a good indication that the user messed up big time. Maybe it would be a good idea to enforce out-of-source builds in a separate PR.

Reply:
Enforcing out-of-source builds is a good idea, but we would definitely need a good user-friendly message.

Reply:
I'm not entirely sure what the CMake-related ignores accomplish either. Following a dev-mode build, and with this gitignore, git still detects the following changes:
```CPackConfig.cmake
CPackSourceConfig.cmake
DartConfiguration.tcl
examples/binding
examples/builders
examples/custom-rtti
examples/dao
examples/inheritance
examples/partials
examples/proxy
examples/serialization
examples/struct-of-arrays
refl-cpp-config-version.cmake
test/refl-cpp-test
```

I think it's best to not attempt to ignore build files explicitly, as doing so correctly would be difficult. Instead, we can start warning users against running cmake in the source root, as @friendlyanon suggested.

Reply:
Some of this comes from the [cpp_starter_project](https://github.com/lefticus/cpp_starter_project) template. There are also some cmake files that hint to user that they should do out-of-source build. But I didn't want to pull these in since you may choose another flavour of how to setup your cmake files. 

Maybe mix and match as you see fit. The idea behind this .gitignore is that users do not commit (by accident) their IDE, OS-specific (i.e. indexing files), or other build files by mistake...

Reply:
I'll give Jason that he knows his C++ well, but that cpp_starter_project is pretty bad when it comes to trivially consumable CMake projects. Such projects require some more thought put into them.